### DISCUSSION TITLE
3D Modle Viewing

### DISCUSSION THREAD

**Discussion by tobiasfalk** [Part-DB/Part-DB-server#805] [Ideas]
Hi, I think it would be nice if one could add a 3D model as "picture" and one then could not only have a static picture but also move it around.

The easiest would most likely be to use STEP files, but having the ability to upload freecad files and maybe the part-db server converts them to step would also be nice.

It would also be nice if one sets a KiCad footprint part-db looks if there is a 3d model for this footprint and uses this. For this, the 3D Model part of the library would most likely be "installed"/download on the server or at least on a place where it is accessible for Part-DB.
Regarding the Std Lib of KiCAD, one must most likely take a detailed look if the licences fit together and look if there are any conflicts/problems.

**Comment by jbtronics** [comment]
The old (legacy) Part-DB version could actually do something like this. Thats why there is a 3D Model option for Footprints under Attachments and there is this repository (https://github.com/Part-DB/PartDB-3D) with 3D models from an old KiCad library version. 
However the library used for 3d models only worked if they were in the pretty obscure x3d format, and I found this feauture not too useful, so I never reimplemented in the new version.

The challenge is that I would like to support multiple 3d model files to view (step would be nice, STL might be useful, and other things). I think the best option for this would be realize it using three.js but that is much more complicated to use than the old 3D viewer library i have used for Part-DB legacy, and i had not the time to look into it yet.

**Reply by tobiasfalk** [reply]
Why not us a freecad instance that is installed on the server to convert the 3dModels to x3d format?

Since FreeCAD can export it https://wiki.freecad.org/Import_Export/de and it should be possible to write a python script with/for freecad that loads a file and exports is as x3d file.

Yes, this requires an installed freecad instance, but this is not too much of a  hurdle since most Linux systems have it available in a repo.
And it would give a rader big list of file formats that are acceptable.

**Comment by jbtronics** [comment]
With the latest commits there is now a builtin 3d model viewer

<img width="1600" height="835" alt="image" src="https://github.com/user-attachments/assets/bbe59750-5873-4afa-afe9-0134419282fe" />

See issue #1507
